### PR TITLE
[admin_style_v3] Products table with variants

### LINKED ISSUE #10694: [BUU] See products in my catalogue
## Context
In this story we are starting to build the actual table and display product data in the page. This is also an opportunity to spike some foundational behaviours for the table itself, especially lateral scrolling when there are too many columns to read the fields comfortably.

Some work has been done in the past on this, and can be seen in `/admin/new_products`. We might decide to build on that or restart.
Issues related to previous work can be found in https://github.com/openfoodfoundation/openfoodnetwork/pull/8996, https://github.com/openfoodfoundation/openfoodnetwork/issues/8930, https://github.com/openfoodfoundation/openfoodnetwork/issues/10350
Some of the above issues are still in progress, once we kick off this card we will archive them so that we can have a cleaner board.

## Out of scope
- display variants
- edit any field
- display columns that are variant specific
- pagination (displayed in the mockups just for future reference)

## Description

**- As an:** enterprise user 
**- On page:** `/admin/products`
**- I want to be able to:** see the products in my catalogue
**- So that:** I can review what I am selling and its stock levels

## Acceptance Criteria & Tests

**Scenario 1: Seeing products - single enterprise**
Given I have one enterprise only
And I have products in my catalogue
And I am in `/admin/products`
And the page is loading my catalogue
When my catalogue has been loaded successfully
Then I see the list of my products ordered by name

**Scenario 2: Seeing products - multiple enterprises**
Given I have multiple enterprises
And I am in `/admin/products`
And the page is loading my catalogue
When my catalogue has been loaded successfully
Then I see the list of the products of all my enterprises ordered by name

## Columns and fields to be displayed

Note: these are in display order

| Column | Display for single enterprise| Display for multiple enterprises | Mandatory | Display style & format |
| --- | --- | --- | --- | --- |
| Name | Yes | Yes | Yes | Semibold, left aligned |
| SKU | Yes | Yes | No | Regular, right aligned |
| Unit type - **EXCLUDE ITEM LABEL** | Yes | Yes | Yes | Regular, right aligned |
| On hand - DISPLAY ON DEMAND IF ALL VARIANTS ARE ON DEMAND | Yes | Yes | Yes | Regular, right aligned |
| Producer | NO | Yes | Yes | Regular, left aligned |
| Category | Yes | Yes | Yes | Regular, left aligned |
| Tax category | Yes | Yes | No | Regular, left aligned |
| Inherits properties - **DISPLAY AS YES / NO** | Yes | Yes | Yes | Regular, left aligned |
| ~~Available on~~ | ~~Yes~~ | ~~Yes~~ | ~~Yes~~ | ~~Regular, right aligned, YYYY-MM-DD~~ |
| ~~TBD - Import date~~ | ~~Yes~~ | ~~Yes~~ | ~~No~~ | ~~Regular, right aligned, YYYY-MM-DD~~ |

## Design specs

The screen for specs is here
https://www.figma.com/file/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?node-id=24%3A645&t=rIk1MTfvkQenGndx-1

A prototype to try the lateral scrolling is here (you need to click and drag on the table to interact)
https://www.figma.com/proto/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?page-id=24%3A642&node-id=24-645&viewport=486%2C678%2C0.64&scaling=min-zoom

**New components and styles**

| Component | Example |
| --- | --- |
| Table header | Screenshot 2023-04-14 at 9 28 39 am |
| Table row with text and relaxed height (we'll build different heights later) | Screenshot 2023-04-14 at 9 26 17 am |
| Scoll bar (might not be needed if browser handles well | Screenshot 2023-04-14 at 9 26 32 am |

### LINKED ISSUE #11005: [BUU] See variants for my products
## Context
In this story we are building on top of https://github.com/openfoodfoundation/openfoodnetwork/issues/10694 to display the variants for each product in the table.

## Out of scope
- edit variants
- add new variants

## Description

**- As an:** enterprise user 
**- On page:** `/admin/products`
**- I want to be able to:** see the variants in my catalogue
**- So that:** I can review what I am selling and its stock levels

## Acceptance Criteria & Tests

**Scenario 1: Seeing variants - single enterprise**
Given I have one enterprise only
And I have products in my catalogue
And I am in `/admin/products`
And the page is loading my catalogue
When my catalogue has been loaded successfully
Then I see the list of my products ordered by name
And within each product I can see its variants

## Columns and fields to be displayed

Note 1: these are in display order, and include the ones that are product specific. The only new column to be added for now is Price.

Note 2: available on and import date don't need to be displayed, as per https://github.com/openfoodfoundation/openfoodnetwork/issues/10694

| Column | Display for single enterprise| Display for multiple enterprises | Mandatory | Display style & format |
| --- | --- | --- | --- | --- |
| Name | Yes | Yes | Yes | Semibold, left aligned |
| SKU | Yes | Yes | No | Regular, right aligned |
| Unit type - **EXCLUDE ITEM LABEL** | Yes | Yes | Yes | Regular, right aligned |
| **Price** | Yes | Yes | Yes | Regular, right aligned |
| On hand - DISPLAY ON DEMAND IF ALL VARIANTS ARE ON DEMAND | Yes | Yes | Yes | Regular, right aligned |
| Producer | NO | Yes | Yes | Regular, left aligned |
| Category | Yes | Yes | Yes | Regular, left aligned |
| Tax category | Yes | Yes | No | Regular, left aligned |
| Inherits properties - **DISPLAY AS YES / NO** | Yes | Yes | Yes | Regular, left aligned |
| ~~Available on~~ | ~~Yes~~ | ~~Yes~~ | ~~Yes~~ | ~~Regular, right aligned, YYYY-MM-DD~~ |
| ~~TBD - Import date~~ | ~~Yes~~ | ~~Yes~~ | ~~No~~ | ~~Regular, right aligned, YYYY-MM-DD~~ |

## Design specs

The screen for specs is here
https://www.figma.com/file/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?type=design&node-id=267%3A832&t=hO5rg4YoMVz0lMCY-1

**New components and styles**
More info on tables is available here https://github.com/openfoodfoundation/openfoodnetwork/wiki/Design-system.

| Component | Example |
| --- | --- |
| Table row with condensed height | Screenshot 2023-06-14 at 11 39 37
|

### PR DESCRIPTION
**Pull request by dacook**
#### What? Why?

- Closes #10694
- Closes #11005
- Work in progress on #10679

Developing the new table for displaying and (future) bulk editing of products.

It's a work in progress, so remaining behind feature toggle. Some of the data values have comments for improvements which will be dealt with as we progress (although feel free to comment on these).

There are some changes outside of the feature toggle, however these don't make user-facing changes.

**Narrow screen, showing clipping (which is a work in progress):**
Screen Shot 2023-06-27 at 3 56 56 pm
**Wide screen, filling full page width:**
Screen Shot 2023-06-27 at 3 56 27 pm

#### What should we test?
No user-facing changes  to test (it will be tested when feature toggle is removed).

The placement of a css class in the admin template is moved, so it's possible this affects the layout of some pages. It affected the spec for "New Order Cycle" page, so it would be worth checking this page appears as expected.

#### Release notes

Changelog Category: 😎 Feature toggle

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Comment by dacook** [comment]
Hmm, this is not rebase-friendly with the [dev branch](https://github.com/openfoodfoundation/openfoodnetwork/pull/11007). 

So I need to rebase that branch on this one first, to ensure it doesn't mess it up and block further development.

**Comment by dacook** [comment]
Flaky spec keeps coming up on this branch.
```
rspec ./spec/system/consumer/caching/shops_caching_spec.rb:22 
```

**Review by jibees — APPROVED** [review]
I'd say it's fine, but I'm a bit concerned of not merging the whole wip branch, since work is very linked ; I've re-looked at the wip branch, and I didn't find anything that should come with this one, so I'd say: good to go! 🤞

File: app/views/admin/products_v3/index.html.haml
Comment (mkllnk) [inline app/views/admin/products_v3/index.html.haml:17]: What happened to your opinion that a page should be progressively enhanced by JS? Shouldn't the products be rendered straight away and only be updated by a reflex?

File: app/webpacker/controllers/productsV3_controller.js
Comment (mkllnk) [inline app/webpacker/controllers/productsV3_controller.js:15]: Maybe it's too early for extraction but I would like to see a more general solution for the "loading spinner. I would expect it to work similar to the `disable-with` attribute on submit buttons. Is there something pre-existing?

I don't know if it would work but we could have a controller inserting the spinner into the the target element when an action is triggered and then it's replaced with new HTML when the target element is re-rendered.

**Review by mkllnk — APPROVED** [review]
Great work! So cool to see this moving forward.

Moving this to Test Ready for a final check.

**Comment by dacook** [comment]
@jibees I've added to this PR:
* Closes https://github.com/openfoodfoundation/openfoodnetwork/issues/10679

File: app/views/admin/products_v3/index.html.haml
Comment (dacook) [reply app/views/admin/products_v3/index.html.haml:17]: Good question! This is how it was specified in #10679.
I have to admit I hadn't thought through this one. while I prepared the PR it was a combination of both JB and I's work which I didn't think to review myself. 

Thinking about it, I agree. Perhaps if it was expensive data to load, we might want to load the rest of the interface first  (ie the search and filter selection) before loading the data. But with pagination that shouldn't be a problem. 
Also in a more general way, we seek to provide a responsive experience with Turbo. 

So.. yeah I'm not sure why we need the loading spinner for the initial load. We do however intend to have it for when you search/filter/paginate.

@jibees, what do you think?

Comment (jibees) [reply app/views/admin/products_v3/index.html.haml:17]: I added a loading spinner as mentionned in the spec. I think it's fine to add such a thing, since once you click on "Search" button, or pagination, interface is blocked by this spinner, and then you cannot click twice. And sometimes yes, it can take some times.

Actually, not sure to follow:
> Shouldn't the products be rendered straight away and only be updated by a reflex?

What do you mean?

File: app/webpacker/controllers/productsV3_controller.js
Comment (jibees) [reply app/webpacker/controllers/productsV3_controller.js:15]: The controller has changed since, don't know if it fits well for you.

It consists of a `loading-controller`, that works with event listener. 

```js
import ApplicationController from "./application_controller";

export default class extends ApplicationController {
  connect() {
    super.connect();
    document.addEventListener("show-loading", this.showLoading);
    document.addEventListener("hide-loading", this.hideLoading);
  }

  disconnect() {
    document.removeEventListener("show-loading", this.showLoading);
    document.removeEventListener("hide-loading", this.hideLoading);
  }

  hideLoading = () => {
    this.element.classList.add("hidden");
  };

  showLoading = () => {
    this.element.classList.remove("hidden");
  };
}
```

and then `products-controller` is simply:

```js
import ApplicationController from "./application_controller";

export default class extends ApplicationController {
  connect() {
    super.connect();
    // Fetch the products on page load
    this.stimulate("Products#fetch");
  }

  beforeReflex() {
    this.showLoading();
  }

  afterReflex() {
    this.hideLoading();
  }

  showLoading = () => {
    const event = new CustomEvent("show-loading");
    document.dispatchEvent(event);
  };

  hideLoading = () => {
    const event = new CustomEvent("hide-loading");
    document.dispatchEvent(event);
  };
}
```

Comment (jibees) [reply app/webpacker/controllers/productsV3_controller.js:15]: Actually, writing my comment, I thought I could do in another (smarter?) way ; See commit 9e5078d59c

File: app/views/admin/products_v3/index.html.haml
Comment (mkllnk) [reply app/views/admin/products_v3/index.html.haml:17]: I mean that the initial page render would not show "Loading" but render the products within the document. The "Loading" would only be used for updates when you sear etc.

File: app/webpacker/controllers/productsV3_controller.js
Comment (mkllnk) [reply app/webpacker/controllers/productsV3_controller.js:15]: That's an interesting way of doing it. I don't have an opinion about it. :shrug:

File: app/views/admin/products_v3/index.html.haml
Comment (jibees) [reply app/views/admin/products_v3/index.html.haml:17]: I personaly prefer to display the "Loading" at startup (as well as updates), just in order to be consistent. But it's not a strong opinion.

**Comment by filipefurtad0** [comment]
Hey @dacook and BUU team :raised_hands: 

I've staged this PR and, after enabling the toggle, I was able to see the page, on the new separator `new_products`, only as superadmin:

image

I see no changes on the products tab:

image

Smoke tested the order cycles page, all seems as before.

My pics differ quite a bit from those you've posted... any idea why? In anycase, as this is a WIP, I'll move to ready-to-go and let you click the merge button ;-)

**Comment by dacook** [comment]
Thanks Filipe!

> My pics differ quite a bit from those you've posted... any idea why?

Sorry for the confusion: "new_products" is the old new products 🤪 . The "new" new products isn't on the menu yet, but should be in the next PR.

**Pull request merged**
